### PR TITLE
Support spmd knn search

### PR DESCRIPTION
**Pull request by ethanglaser**
## Description

Fix errors that occur when calling kNN search (ie without responses result_option set)

Corresponding sklearnex PR adding these interfaces to API: TODO

Checklist to comply with **before moving PR from draft**:

**PR completeness and readability**

- [ ] I have reviewed my changes thoroughly before submitting this pull request.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [ ] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [ ] I have added a respective label(s) to PR if I have a permission for that.
- [ ] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [ ] I have run it locally and tested the changes extensively.
- [ ] All CI jobs are green or I have provided justification why they aren't.
- [ ] I have extended testing suite if new functionality was introduced in this PR.

### CODE REVIEW

File: cpp/oneapi/dal/algo/knn/backend/gpu/infer_kernel_impl_dpc_distr.hpp [hunk 1]
```diff
@@ -206,25 +206,30 @@ class knn_callback_distr {
...
         auto min_indc_dest = indices_.get_row_slice(first, last);
-        auto min_resp_dest = intermediate_responses_.get_row_slice(first, last);
 
         // add global offset value to input indices
         ONEDAL_ASSERT(global_index_offset_ != -1);
-        auto treat_event = pr::treat_indices(queue_,
-                                             inp_indices,
-                                             global_index_offset_,
-                                             { copy_current_resp_event });
+        bk::event_vector ndeps = deps;
```
Comment (Alexandr-Solovev) [inline cpp/oneapi/dal/algo/knn/backend/gpu/infer_kernel_impl_dpc_distr.hpp:230]: does it make sense to reuse deps with?
`deps.push_back(copy_current_resp_event);`

File: cpp/oneapi/dal/algo/knn/backend/gpu/infer_kernel_impl_dpc_distr.hpp [hunk 2]
```diff
@@ -265,27 +271,33 @@ class knn_callback_distr {
...
             if (this->compute_sqrt_) {
                 copy_sqrt_event =
                     copy_with_sqrt(queue_, min_dist_dest, min_dist_dest, { select_indc_event });
             }
-            auto final_event = this->output_responses(bounds,
-                                                      indices_,
-                                                      distances_,
-                                                      { select_indc_event, copy_sqrt_event });
-            return final_event;
+            if (result_options_.test(result_options::responses)) {
```
Comment (Alexandr-Solovev) [inline cpp/oneapi/dal/algo/knn/backend/gpu/infer_kernel_impl_dpc_distr.hpp:292]: you have 2 returns in the same if
`                return this->output_responses(bounds,
                                              indices_,
                                              distances_,
                                              { select_indc_event, copy_sqrt_event });`
                 and `            if (this->compute_sqrt_) {
                return copy_sqrt_event;
            }`           
      Can they conflict?

File: cpp/oneapi/dal/algo/knn/backend/gpu/infer_kernel_impl_dpc_distr.hpp (hunk 1 above)
Comment (ethanglaser) [reply cpp/oneapi/dal/algo/knn/backend/gpu/infer_kernel_impl_dpc_distr.hpp:230]: Yes, not sure what I was trying to do here haha

File: cpp/oneapi/dal/algo/knn/backend/gpu/infer_kernel_impl_dpc_distr.hpp (hunk 2 above)
Comment (ethanglaser) [reply cpp/oneapi/dal/algo/knn/backend/gpu/infer_kernel_impl_dpc_distr.hpp:292]: There are 3 cases with 3 return possibilities. If one returns, then the others will not be reached.

**Comment by Alexandr-Solovev** [comment]
In file included from cpp/oneapi/dal/algo/knn/backend/gpu/infer_kernel_brute_force_dpc.cpp:24:
./cpp/oneapi/dal/algo/knn/backend/gpu/infer_kernel_impl_dpc_distr.hpp:230:14: error: no matching member function for call to 'push_back'
  230 |         deps.push_back(copy_current_resp_event);
      |         ~~~~~^~~~~~~~~
/usr/lib/gcc/x86_64-linux-gnu/14/../../../../include/c++/14/bits/stl_vector.h:1283:7: note: candidate function not viable: 'this' argument has type 'const bk::event_vector' (aka 'const vector<sycl::event>'), but method is not marked const
 1283 |       push_back(const value_type& __x)
      |       ^
/usr/lib/gcc/x86_64-linux-gnu/14/../../../../include/c++/14/bits/stl_vector.h:1300:7: note: candidate function not viable: 'this' argument has type 'const bk::event_vector' (aka 'const vector<sycl::event>'), but method is not marked const

File: cpp/oneapi/dal/algo/knn/backend/gpu/infer_kernel_impl_dpc_distr.hpp (hunk 1 above)
Comment (ethanglaser) [reply cpp/oneapi/dal/algo/knn/backend/gpu/infer_kernel_impl_dpc_distr.hpp:230]: Update: this was causing failure, so it does make sense to keep as is

**Review by Alexandr-Solovev — APPROVED** [review]
